File: src/main/java/com/example/reactive/chapter1/completable/CompletableFutureShoppingCartService.java
package com.example.reactive.chapter1.completable;

import java.util.concurrent.CompletableFuture;

import com.example.reactive.chapter1.commons.Input;
import com.example.reactive.chapter1.commons.Output;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@RequiredArgsConstructor
public class CompletableFutureShoppingCartService implements ShoppingCartService {

    @Override
    public CompletableFuture<Output> calculate(Input input) {

        log.info("CompletableFuture Style calculating...");

        return CompletableFuture.supplyAsync(() -> {
            try {
                Thread.sleep(2000);

            } catch (InterruptedException e) {
                log.error(e.getMessage(), e);
            }
            return new Output();
        });
    }
}

File: src/main/kotlin/com/example/reactor/store/completable/CompletableFutureShoppingCartService.kt
package com.example.reactor.store.completable

import com.example.reactor.shared.logger
import com.example.reactor.store.domain.Input
import com.example.reactor.store.domain.Output
import java.util.concurrent.CompletableFuture

class CompletableFutureShoppingCartService {
    fun calculate(input: Input): CompletableFuture<Output> {
        return CompletableFuture.supplyAsync {
            Thread.sleep(2_000)
            logger().info("calculated!")
            Output()
        }
    }
}
